log/scope_test.go

```diff
@@ -218,6 +220,130 @@ func TestBasicScopes(t *testing.T) {
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").AddLabel("baz", 123).AddLabel("qux", 0.123).Debuga("Hello")
+		s.Debuga("Hello")
+		_ = Sync()
+	})
+
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	expect := `Hello {"foo":"bar","baz":1,"qux":0.123}`
+	mustRegexMatchString(t, lines[0], expect)
+	mustRegexMatchString(t, lines[1], expect)
+}
+
+func TestScopeWithLabel(t *testing.T) {
+	const name = "TestScope"
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").WithLabel("baz", 123).Debuga("Hello")
+		s.Debuga("Hello")
+		_ = Sync()
+	})
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello {"foo":"bar","baz":123}`)
+	mustRegexMatchString(t, lines[1], `Hello {"foo":"bar"}`)
+}
+
+func TestScopeClearLabel(t *testing.T) {
+	const name = "TestScope"
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").AddLabel("baz", 123).AddLabel("qux", 0.123)
+		s.ClearLabelKey("qux")
+		s.Debuga("Hello")
+		s.ClearAllLabels()
+		s.Debuga("Hello")
+
+		_ = Sync()
+	})
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello {"foo":"bar","baz":123}`)
+	mustRegexMatchString(t, lines[1], `Hello$`)
+}
+
+func TestScopeLocal(t *testing.T) {
+	const name = "TestScope"
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").AddLabel("baz", 123)
+		s2 := s.Local()
+		s2.ClearAllLabels()
+		s.Debuga("Hello")
+		s2.Debuga("Hello")
+
+		_ = Sync()
+	})
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello {"foo":"bar","baz":123}`)
+	mustRegexMatchString(t, lines[1], `Hello$`)
+}
+
+func TestScopeErrorDictionary(t *testing.T) {
+	const name = "TestScope"
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	ie := &errdict.IstioErrorStruct{
+		MoreInfo:    "MoreInfo",
+		Impact:      "Impact",
+		Action:      "Action",
+		LikelyCause: "LikelyCause",
+	}
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.WithLabel("foo", "bar").Debuga(ie, "Hello")
+
+		_ = Sync()
+	})
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `{"moreInfo":"MoreInfo","impact":"Impact","action":"Action","likelyCauses":"LikelyCause","foo":"bar","message":"Hello"}`)
```

Comment:
i'm a little nervous that the output changes format based on the inclusion of the `ie` in the call to `Debuga()`. That seems like it might surprise.

Somewhat related, I don't know what the right thing here is, but I worry that we'll have alternating log lines like:

```text
Hello {"foo": "bar", "baz": "goo"}
{"moreInfo"... "message":"Hello"}
Goodbye {"foo"...}
{"moreInfo"..."message":"Goodbye"}
```

This seems potentially like a mess to parse. Am I reading this incorrectly?

Reply:
Good point, how about:
```
Hello {"foo": "bar", "baz": "goo"}
Hello { "message":"Hello","moreInfo"...}
```

Reply:
i'm fairly torn on this. what does zap default to?  does the console encoder thingy use half text, half JSON?  if so, then we should probably match that, and then only switch to full JSON if that's requested in the config (which I believe is your proposal).

As an aside, I almost feel like cli-style (`Hello foo=bar baz=goo`) or full JSON (`{"message":"Hello", "foo":"bar", ...}`) are better than mixing JSON and free text, but that seems like an issue for zap, not Istio.